mt7622: add support for Unielec U7622-02 V2.1

hashyuan opened the pull request:
Hardware specification

----------------------
* MT7622B SoC, 2 x ARM Cortex-A53 1.35GHz
* 512MB/1GB DDR3
* 16MB NOR Flash
* 8GB EMMC
* 4 x 4 2.4G WIFI up to 800Mbps
* 5 x 10/100/1000 ethernet ports (RTL8367S)
* 1 x SD card slot
* 2 x M.2 for 4G LTE/5G NG modules
* 2 x SIM slots
* 1 x miniPCIE
* 1 x SATA 3.0
* 10 x LEDs
* J6: debug UART
* J8: 1 x UART
       1 x I2C
       1 x SPI


WIFI calibration section:
----------------------
NOR：0xc0000 - 0xe0000
EMMC：0x500000 - 0x600000


Installation using SD card
----------------------
gzip -d openwrt-mediatek-mt7622-unielec_u7622-02-sdcard.img.gz
sudo dd if=openwrt-mediatek-mt7622-unielec_u7622-02-sdcard.img of=/dev/sdb

After booting from SD card, choose the boot options

>        ( ( ( OpenWrt-SD card) ) )
>
>      1. Run default boot command.
>      2. Boot production system from SD card.
>      3. Boot recovery system from SD card.
>      4. Install bootloader, recovery and production to eMMC.
>      5. Install bootloader, production to NOR Flash.
>      6. Reboot.
>      0. U-Boot console


Using SATA3.0
----------------------
setenv use_sata 1
saveenv


Signed-off-by: Hash Yuan <hashyuan@gmail.com>

hashyuan committed:
mt7622: add support for Unielec U7622-02 V2.1
Hardware specification

----------------------
* MT7622B SoC, 2 x ARM Cortex-A53 1.35GHz
* 512MB/1GB DDR3
* 16MB NOR Flash
* 8GB EMMC
* 4 x 4 2.4G WIFI up to 800Mbps
* 5 x 10/100/1000 ethernet ports (RTL8367S)
* 1 x SD card slot
* 2 x M.2 for 4G LTE/5G NG modules
* 2 x SIM slots
* 1 x miniPCIE
* 1 x SATA 3.0
* 10 x LEDs
* J6: debug UART
* J8: 1 x UART
       1 x I2C
       1 x SPI

WIFI calibration section:
----------------------
NOR：0xc0000 - 0xe0000
EMMC：0x500000 - 0x600000

Installation using SD card
----------------------
gzip -d openwrt-mediatek-mt7622-unielec_u7622-02-sdcard.img.gz
sudo dd if=openwrt-mediatek-mt7622-unielec_u7622-02-sdcard.img of=/dev/sdb

After booting from SD card, choose the boot options

>        ( ( ( OpenWrt-SD card) ) )
>
>      1. Run default boot command.
>      2. Boot production system from SD card.
>      3. Boot recovery system from SD card.
>      4. Install bootloader, recovery and production to eMMC.
>      5. Install bootloader, production to NOR Flash.
>      6. Reboot.
>      0. U-Boot console

Using SATA3.0
----------------------
setenv use_sata 1
saveenv

Signed-off-by: Hash Yuan <hashyuan@gmail.com>

Ansuel commented:
@hashyuan missing installation instruction ?

mrkiko commented:
And maybe recovery instructions if applicable.

mrkiko commented:
The hardware seems very very nice. thanks a lot for your work!
And sorry for the rudeness of the previous message.

Enrico

Ansuel commented on target/linux/mediatek/dts/mt7622-unielec-u7622-02.dts:
```diff
@@ -0,0 +1,549 @@
+#include "mt7622.dtsi"
+#include "mt6380.dtsi"
+
+//#define USE_PCIE1
```
Mhhh why? is there a variant of this device that use pcie1?

hashyuan replied:
> Mhhh why? is there a variant of this device that use pcie1?

U7622-02 use 4 capacitors to switch between PCIE1 and SATA

Ansuel replied:
So again what is the default state and how to transition from the 2 state?

Do you have to do hw modification to change this? GPIO? a hw switch on the board?

We can't have these kind of commented define and ifdef in dts. To support both of them you should make a dtsi and produce 2 different dts and 2 variant of this device.

hashyuan replied:
Customers define the default state. Unlike banana-pi-r64 using GPIO to switch between PCIE1 and SATA, U7622-02 use 4 capacitors to switch, so you have to to move the 4 capacitors to decide the signals connected to M.2 PCIE or SATA3.0.

hashyuan replied:
Alright, I will add a uboot env option "use_sata" to decide it. During booting kernel, a uboot script will change the kernel dts according to the use_sata uboot env value.

Ansuel replied:
@hashyuan ok it's not the best solution but it's a solution :D

I would ask you to add some comments in the dts about the related nodes 
and improve the documentation about use_sata

Something like
```
U7622-02 use 4 capacitors to switch, so you have to to move the 4 capacitors to decide the signals connected to M.2 PCIE or SATA3.0.
Uboot will patch the dts to enable and disable the related nodes.
To make the switch set use_sata in the uboot env........
```

This is just an example but it's really to not lose info about this process and the option after this gets merged. (on openwrt we check the commit but sometime we lose all the discussion done on the pr/mailing list so better put all the info in the commit description)

hashyuan replied:
A comment is added.

hashyuan commented:
> @hashyuan missing installation instruction ?

updated

Ansuel commented:
@hashyuan Thanks for adding installation instruction

hashyuan committed:
mt7622: Update mt7622-unielec-u7622-02.dts

Now the default state is PCE1

Signed-off-by: Hash Yuan <hashyuan@gmail.com>

hashyuan committed:
mt7622: Update mt7622-unielec-u7622-02.dts

The default state is M.2 PCIE1, now add a uboot env use_sata option
to let uboot script change the kernel dts to use SATA3.0

Signed-off-by: Hash Yuan <hashyuan@gmail.com>

Ansuel commented on target/linux/mediatek/dts/mt7622-unielec-u7622-02.dts:
```diff
@@ -0,0 +1,533 @@
+		compatible = "gpio-leds";
+
+		M2-bottom {
+			label = "M2-bottom";
```
you should use the standard way to define leds with `color:use`

hashyuan replied:
updated

Ansuel commented on target/linux/mediatek/dts/mt7622-unielec-u7622-02.dts:
```diff
@@ -0,0 +1,533 @@
+		};
+
+		M2-up {
+			label = "M2-up";
```
you should use the standard way to define leds with `color:use`

hashyuan replied:
updated

hashyuan committed:
mt7622: Update mt7622-unielec-u7622-02.dts

add color green to the gpio led label

Signed-off-by: Hash Yuan <hashyuan@gmail.com>

hashyuan committed:
mt7622: Update mt7622-unielec-u7622-02.dts

add a comment about a uboot env variable use_sata.

Signed-off-by: Hash Yuan <hashyuan@gmail.com>

hashyuan commented:
@Ansuel ，Is it OK to close this pull and make another pull with the changes to fix the Test Formalities failure?

Ansuel commented:
Why? Can't you just force push?

Il Lun 1 Ago 2022, 04:24 hashyuan ***@***.***> ha scritto:

> @Ansuel <https://github.com/Ansuel> ，Is it OK to close this pull and make
> another pull with the changes to fix the Test Formalities failure?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/10283#issuecomment-1200612522>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQQDDAFGMKXS2EY7ME3VW4YOFANCNFSM53VLOKLQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hashyuan commented:
> Why? Can't you just force push? Il Lun 1 Ago 2022, 04:24 hashyuan ***@***.***> ha scritto:
> […](#)
> @Ansuel <https://github.com/Ansuel> ，Is it OK to close this pull and make another pull with the changes to fix the Test Formalities failure? — Reply to this email directly, view it on GitHub <[#10283 (comment)](https://github.com/openwrt/openwrt/pull/10283#issuecomment-1200612522)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AE2ZMQQDDAFGMKXS2EY7ME3VW4YOFANCNFSM53VLOKLQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

hashyuan commented:
I accidently clicked the Close button. resume it.

hashyuan commented:
> Why? Can't you just force push? Il Lun 1 Ago 2022, 04:24 hashyuan ***@***.***> ha scritto:
> […](#)
> @Ansuel <https://github.com/Ansuel> ，Is it OK to close this pull and make another pull with the changes to fix the Test Formalities failure? — Reply to this email directly, view it on GitHub <[#10283 (comment)](https://github.com/openwrt/openwrt/pull/10283#issuecomment-1200612522)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AE2ZMQQDDAFGMKXS2EY7ME3VW4YOFANCNFSM53VLOKLQ> . You are receiving this because you were mentioned.Message ID: ***@***.***>

Closing the pull and make another pull is the easy way I thought. 
It takes a while for me to make the changes to pass the Test Formalities.

hauke commented on package/boot/uboot-mediatek/patches/480-add-unielec-u7622-02.patch:
```diff
@@ -0,0 +1,2188 @@
+-		return ret;
+-	return fdtdec_setup_mem_size_base();
++	gd->ram_size = get_ram_size((long *)CONFIG_SYS_SDRAM_BASE,
++					SZ_2G);
```
Why is this change needed?

hashyuan replied:
Unielec U7622-02 V2.1 has 512MB/1GB memory configuration. I don't want to add another *-1GB.dts file to support 1GB memory. The change should work on the other mt7622 devices.

hauke replied:
Please also try to get such changes into the upstream kernel. We would like to reduce  the number of patches in OpenWrt.

dangowrt replied:
Afaik you should have only one DTS for the 1GB variant, U-Boot will detect if there are only 512MB and set the correct values in the device tree when starting Linux.

hauke commented on package/boot/uboot-mediatek/patches/480-add-unielec-u7622-02.patch:
```diff
@@ -0,0 +1,2188 @@
+ 
+ void reset_cpu(void)
+--- a/configs/mt7622_unielec_u7622-02-emmc_defconfig
++++ b/configs/mt7622_unielec_u7622-02-emmc_defconfig
```
The defconfig is not needed here. If you create defconfigs please use the `make savedefconfig ` from the kernel, it will minimize it.

hashyuan replied:
I don't see why the emmc_defconfig is not needed here.
I don't know the `make savedefconfig` can reduce the config size, Thanks. I will update the nor_defconfig.

hauke replied:
The defconfig is not used by OpenWrt, but we have to maintain this patch. I would prefer if you do not add these kernel defconfig files.

hashyuan committed:
mt7622: add an echo to satisfy the uboot script && check

when run boot_production and use_sata is 0, the && check always fail, the bootm $loadaddr will never execute.

Signed-off-by: Hash Yuan <hashyuan@gmail.com>

hashyuan committed:
mt7622: reduce mt7622_unielec_u7622-02-nor_defconfig size

replaced with make savedefconfig generated defconfig.

Signed-off-by: Hash Yuan <hashyuan@gmail.com>
